packages/framer-motion/src/value/types/complex/index.ts

```diff
@@ -18,107 +13,95 @@ function test(v: any) {
     )
 }
 
-export interface ComplexValueInfo {
-    value: string
-    values: Array<CSSVariableToken | Color | number>
-    numVars: number
-    numColors: number
-    numNumbers: number
-    tokenised: string
-}
-
-interface Tokeniser {
-    regex: RegExp
-    countKey: string
-    token: string
-    parse: (value: string) => any
-}
+const NUMBER_TOKEN = "number"
+const COLOR_TOKEN = "color"
+const VAR_TOKEN = "var"
+const VAR_FUNCTION_TOKEN = "var("
+const SPLIT_TOKEN = "${}"
 
-const cssVarTokeniser: Tokeniser = {
-    regex: cssVariableRegex,
-    countKey: "Vars",
-    token: "${v}",
-    parse: noop,
-}
+export type ComplexValues = Array<CSSVariableToken | string | number | Color>
 
-const colorTokeniser: Tokeniser = {
-    regex: colorRegex,
-    countKey: "Colors",
-    token: "${c}",
-    parse: color.parse,
+export type ValueIndexes = {
+    color: number[]
+    number: number[]
+    var: number[]
 }
 
-const numberTokeniser: Tokeniser = {
-    regex: floatRegex,
-    countKey: "Numbers",
-    token: "${n}",
-    parse: number.parse,
+export interface ComplexValueInfo {
+    values: ComplexValues
+    split: string[]
+    indexes: ValueIndexes
+    types: Array<keyof ValueIndexes>
 }
 
-function tokenise(
-    info: ComplexValueInfo,
-    { regex, countKey, token, parse }: Tokeniser
-) {
-    const matches = info.tokenised.match(regex)
-
-    if (!matches) return
-
-    info["num" + countKey] = matches.length
-    info.tokenised = info.tokenised.replace(regex, token)
-    info.values.push(...(matches.map(parse) as any))
-}
+const complexRegex =
+    /(var\s*\(\s*--[\w-]+(\s*,\s*(?:(?:[^)(]|\((?:[^)(]+|\([^)(]*\))*\))*)+)?\s*\))|(#[0-9a-f]{3,8}|(rgb|hsl)a?\((-?[\d\.]+%?[,\s]+){2}(-?[\d\.]+%?)\s*[\,\/]?\s*[\d\.]*%?\))|((-)?([\d]*\.?[\d])+)/gi
```

Comment:
🤯

Reply:
Ha its not as clever as it looks, just three capture groups of preexisting color/number/var() patterns